src/core/include/lattice/hal/default/dcrtpoly-impl.h

```diff
@@ -669,30 +669,36 @@ void DCRTPolyImpl<VecType>::SetValuesToZero() {
 }
 
 template <typename VecType>
-DCRTPolyImpl<VecType> DCRTPolyImpl<VecType>::SetValuesModSwitch(const DCRTPolyImpl& element, NativeInteger modulus) {
+void DCRTPolyImpl<VecType>::SetValuesModSwitch(const DCRTPolyImpl& element, const NativeInteger& modulus) {
     if (element.GetNumOfElements() != 1) {
         OPENFHE_THROW(not_implemented_error, "SetValuesModSwitch is implemented only for a DCRTPoly with one tower.");
     }
 
-    auto Q = element.GetModulus();
+    auto Q             = element.GetModulus();
+    double Qmod_double = Q.ConvertToDouble() / modulus.ConvertToDouble();
     this->m_params->SetOriginalModulus(modulus);
 
-    for (size_t i = 0; i < m_vectors.size(); i++) {
-        NativeVector temp(this->GetRingDimension());
-        temp.SetModulus(modulus);
-        auto input = element.Clone();
-        input.SetFormat(Format::COEFFICIENT);
+    auto input{element.GetElementAtIndex(0)};
+    input.SetFormat(Format::COEFFICIENT);
+
+    size_t size{m_vectors.size()};
+    size_t N(m_params->GetRingDimension());
+
+    if (N > this->GetRingDimension())
```

Comment:
(N = m_params->GetRingDimension()) and this->GetRingDimension() are the same value. Did you mean to check against element.m_params->GetRingDimension()?

Reply:
Thanks for the observation. I changed it to:
    size_t N_elem(element.m_params->GetRingDimension());
    size_t N(this->GetRingDimension())

    if (N_elem > N)